components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,420 @@ package controllers
+		log:            logger,
+		directorClient: directorClient,
+		webhookClient:  webhookClient,
+	}
 }
 
 // +kubebuilder:rbac:groups=operations.compass,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operations.compass,resources=operations/status,verbs=get;update;patch
 
 func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+	ctx := context.Background()
+	logger := r.log.WithValues("operation", req.NamespacedName)
+
+	operation, err := r.k8sClient.Get(ctx, req.NamespacedName)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Unable to retrieve %s from API server", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	requestData, err := parseRequestData(operation)
+	if err != nil {
+		logger.Error(err, operationError("Unable to parse request data", operation))
+		return ctrl.Result{}, nil
+	}
+
+	ctx = tenant.SaveToContext(ctx, requestData.TenantID)
+	app, err := r.directorClient.FetchApplication(ctx, operation.Spec.ResourceID)
+	if err != nil {
+		logger.Error(err, operationError("Unable to fetch application", operation))
+
+		if isReconciliationTimeoutReached(operation, time.Duration(r.webhookConfig.TimeoutFactor)*r.webhookConfig.ReconciliationTimeout) {
+			if err := r.k8sClient.Delete(ctx, operation); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{RequeueAfter: r.webhookConfig.RequeueInterval}, nil
+	}
+
+	operation = setDefaultStatus(*operation)
+
+	if app.Result.Ready {
+		operation = setConditionStatus(*operation, v1alpha1.ConditionTypeReady, corev1.ConditionTrue, "")
+		operation.Status.Phase = v1alpha1.StateSuccess
+		if app.Result.Error != nil && *app.Result.Error != "" {
+			operation = setConditionStatus(*operation, v1alpha1.ConditionTypeError, corev1.ConditionTrue, *app.Result.Error)
+			operation.Status.Phase = v1alpha1.StateFailed
+		}
+		if err := r.k8sClient.UpdateStatus(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if app.Result.Error != nil && *app.Result.Error != "" {
+		operation = setErrorStatus(*operation, *app.Result.Error)
+		operation.Status.Phase = v1alpha1.StateFailed
+		if err := r.k8sClient.UpdateStatus(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	webhooks, err := retrieveWebhooks(app.Result.Webhooks, operation.Spec.WebhookIDs)
+	if err != nil {
+		logger.Error(err, operationError("Unable to retrieve webhooks", operation))
+		return ctrl.Result{}, nil
+	}
+
+	if isReconciliationTimeoutReached(operation, calculateReconciliationTimeout(webhooks, r.webhookConfig.WebhookTimeout)) {
+		opErr := errors.New("reconciliation timeout reached")
+		return r.updateStatusWithError(ctx, logger, operation, opErr)
+	}
+
+	webhookEntity := webhooks[operation.Spec.WebhookIDs[0]]
```

Comment:
Initially the idea was to try to make the controller generic and lay the foundation for multiple webhook execution in the future. Eventually however webhook validation logic was added to ensure that only 1 webhook is expected to be executed and also the 0-indexing came along. This obviously causes conflict with the original premise to have the controller generic.
At this point in time I think it might be wiser to limit the controller to a single webhook and eliminate the confusion with multiple webhooks for the sake of more robust controller execution (ensuring always that the input operation will try to execute only a single webhook). Even though the Director component ensures that a single webhook will be placed in the `WebhookIDs` slice, I think it's important for the Operations Controller to also validate this as someone might have accidentally (or not) updated the Operation CR at some point and put the resource in an invalid state which might break the controller execution eventually.